src/io/flutter/run/FlutterReloadManager.java

```diff
@@ -0,0 +1,192 @@
+      // Check for syntax errors in the file - we'll reload after they fix and save.
+      if (documentHasErrors(editor.getDocument())) {
+        return;
+      }
+    }
+
+    // Reload if there are any unsaved files, or if there were any changed project
+    // files since the last reload.
+    if (hadUnsavedDocuments || app.hasChangesSinceLastReload()) {
+      if (editor != null) {
+        showNotification(editor, "Performing hot reload…", false);
+      }
+
+      app.performHotReload(supportsPauseAfterReload()).thenAccept(result -> {
+        if (result.ok()) {
+          return;
+        }
+
+        if (editor != null) {
+          showNotification(editor, result.getMessage(), true);
+        }
+        else {
+          // TODO(devoncarew): Improve flutter_tools reload error messages.
+          showRunNotificationError(app, "Hot Reload", result.getMessage());
+        }
+      });
+    }
+  }
+
+  public void saveAllAndReload(@NotNull FlutterApp app) {
+    if (app.isStarted()) {
+      FileDocumentManager.getInstance().saveAllDocuments();
+      app.performHotReload(supportsPauseAfterReload()).thenAccept(result -> {
+        if (!result.ok()) {
+          showRunNotificationError(app, "Hot Reload", result.getMessage());
+        }
+      });
+    }
+  }
+
+  public void saveAllAndRestart(@NotNull FlutterApp app) {
+    if (app.isStarted()) {
+      FileDocumentManager.getInstance().saveAllDocuments();
+      app.performRestartApp().thenAccept(result -> {
+        if (!result.ok()) {
+          showRunNotificationError(app, "Full Restart", result.getMessage());
+        }
+      });
+    }
+  }
+
+  private FlutterApp getApp(AnAction reloadAction) {
+    if (reloadAction instanceof FlutterAppAction) {
+      return ((FlutterAppAction)reloadAction).getApp();
+    }
+
+    return null;
+  }
+
+  private void showRunNotificationError(FlutterApp app, String title, String message) {
+    final String toolWindowId = app.getMode() == RunMode.RUN ? ToolWindowId.RUN : ToolWindowId.DEBUG;
+    final NotificationGroup notificationGroup =
+      NotificationGroup.toolWindowGroup(FlutterRunNotifications.GROUP_DISPLAY_ID, toolWindowId, false);
+
+    final Notification notification = notificationGroup.createNotification(title, message, NotificationType.ERROR, null);
+    notification.setIcon(FlutterIcons.Flutter);
+    notification.notify(myProject);
+  }
+
+  private FlutterApp getApp() {
+    final AnAction action = ActionManager.getInstance().getAction(ReloadFlutterApp.ID);
+    return action instanceof FlutterAppAction ? ((FlutterAppAction)action).getApp() : null;
+  }
+
+  private boolean supportsPauseAfterReload() {
+    return DartPluginCapabilities.isSupported("supports.pausePostRequest");
+  }
+
+  // TODO(devoncarew): This technique still allows some syntax errors through; can we
+  // improve the IntelliJ grammer? Are there other issues in the AST we should look for?
+  private boolean documentHasErrors(@NotNull Document document) {
```

Comment:
I'm curious why you don't just analyze the content of the analyzer errors view. Is it due to the complexity and timing of update files > analyze files > update error view?